File: C/sort/quick_sort.c
//QUICK SORT WORKS ON THE BASIS OF DIVIDE-AND-CONQUER ALGORITHM.THIS ALGORITHM IS FASTER THAN OTHER SORTING ALGORITHMS.
//TIME COMPLEXITY : O(n log n)
//SPACE COMPLEXITY :O(log n)

#include <stdio.h>

#include <conio.h>

#define capacity 100
int partition(int array[], int beggining, int end);
void quick_sort(int array[], int beggining, int end);
int partition(int array[], int beggining, int end) {
    int low, high, temp, location, flag;
    location = low = beggining;
    high = end;
    flag = 0;
    while (flag != 1) {
        while ((array[location] <= array[high]) && (location != high))
            high--;
        if (location == high)
            flag = 1;
        else if (array[location] > array[high]) {
            temp = array[location];
            array[location] = array[high];
            array[high] = temp;
            location = high;
        }
        if (flag != 1) {
            while ((array[location] >= array[low]) && (location != low))
                low++;
            if (location == low)
                flag = 1;
            else if (array[location] < array[low]) {
                temp = array[location];
                array[location] = array[low];
                array[low] = temp;
                location = low;
            }
        }
    }
    return location;
}
void quick_sort(int array[], int beggining, int end) {
    int location;
    if (beggining < end) {
        location = partition(array, beggining, end);
        quick_sort(array, beggining, location - 1);
        quick_sort(array, location + 1, end);
    }
}
void main() {
    int array[capacity], i, number_of_elements;
    printf("\n Enter the number of elements in the array: ");
    scanf("%d", & number_of_elements);
    printf("Enter the elements of the array: \n");
    for (i = 0; i < number_of_elements; i++) {
        scanf("%d", & array[i]);
    }
    quick_sort(array, 0, number_of_elements - 1);
    printf("\n The sorted array is: \n");
    for (i = 0; i < number_of_elements; i++)
        printf(" %d\t", array[i]);
    getch();
}

//OUTPUT:
//Enter the number of elements in the array: 5
//Enter the elements of the array:
//12      1       34      3       2
//The sorted array is:
//1       2       3       12      34

File: Julia/sort/quick_sort.jl
"""Julia program to implement Quick sort
Quick Sort is a Divide and Conquer strategy based sorting algoithm, that partitions the array
around a selected pivot element. In the below implemented version of Quick Sort we always took the 
last element as the pivot element.

"""
function  partition(arr, low, high)
    # Select the last element as pivot
    pivot = arr[high]
    i = low - 1
    for j in low:high
        # Check if there is any element less than the pivot element
        if(arr[j] < pivot)
            i = i + 1
            # Swap such an element with the to the start position
            arr[i], arr[j] = arr[j], arr[i]
        end
    end
    # Now place the pivot to its original position
    arr[i+1], arr[high] = arr[high], arr[i+1]
    return (i+1)
end

function quick_sort(arr, low, high)
    if (low < high)
        pi = partition(arr, low, high)
        # Recursively call quick_sort function
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)
    end
end
       


print("How many numbers do you want to sort? ")
n = readline()
n = parse(Int, n)
if (n <= 0)
    println("No numbers to sort!!!")
    exit()
end
arr = Int[]
print("Enter the numbers: ")
arr = [parse(Int, num) for num in split(readline())]
quick_sort(arr, 1, n)
print("The numbers in sorted order is ")
print(arr)



"""
Time Complexity - O(n^2), where 'n' is the size of the array
Space Complexity - O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE I
How many numbers do you want to sort? 5
Enter the numbers: 4 2 5 1 3
The numbers in sorted order is [1, 2, 3, 4, 5]

"""
